File: src/features/work-hours/domain/use-cases/get-work-entries.ts
import type { WorkEntry } from "@/features/work-hours/domain/entities/work-hours";
import type { WorkEntryRepository } from "@/features/work-hours/domain/repositories/work-entry-repository";

export type GetWorkEntriesUseCase = (
  presenceId: string,
) => Promise<WorkEntry[]>;

export function makeGetWorkEntriesUseCase(
  repo: WorkEntryRepository,
): GetWorkEntriesUseCase {
  return (presenceId) => repo.listByPresence(presenceId);
}

File: src-tauri/src/application/use_cases/get_work_entries.rs
use std::sync::Arc;

use crate::application::dto::work_entry_dto::WorkEntryDto;
use crate::domain::error::DomainError;
use crate::domain::repositories::work_entry_repository::WorkEntryRepository;

/// Fetch the work entries of a presence day, to fill the hours page.
pub struct GetWorkEntriesUseCase {
    entries: Arc<dyn WorkEntryRepository>,
}

impl GetWorkEntriesUseCase {
    pub fn new(entries: Arc<dyn WorkEntryRepository>) -> Self {
        Self { entries }
    }

    pub async fn execute(&self, presence_id: &str) -> Result<Vec<WorkEntryDto>, DomainError> {
        let presence_id = presence_id.trim();
        if presence_id.is_empty() {
            return Err(DomainError::Validation(
                "presenceId is required".to_string(),
            ));
        }
        let entries = self.entries.list_by_presence(presence_id).await?;
        Ok(entries.into_iter().map(WorkEntryDto::from).collect())
    }
}
